File: services/shared/audit_logger.py
import boto3
import json
import logging
import uuid
from datetime import datetime
from typing import Dict, Any, Optional
from botocore.exceptions import ClientError

logger = logging.getLogger(__name__)

class AuditLogger:
    """
    Shared library for sending audit log messages to SQS queue.
    
    This library should be used by all services that need to log CRUD operations
    for compliance and audit purposes.
    """
    
    def __init__(self, queue_url: str, source_service: str):
        """
        Initialize the audit logger.
        
        Args:
            queue_url: SQS queue URL for audit logs
            source_service: Name of the service using this logger
        """
        self.sqs = boto3.client('sqs')
        self.queue_url = queue_url
        self.source_service = source_service
        
    def log_create(self, client_id: str, agent_id: str, after_value: str, 
                   correlation_id: Optional[str] = None) -> bool:
        """
        Log a CREATE operation.
        
        Args:
            client_id: ID of the client
            agent_id: ID of the agent performing the operation
            after_value: The value created (e.g., "LEE|ABC")
            correlation_id: Optional correlation ID for request tracing
            
        Returns:
            True if log was sent successfully, False otherwise
        """
        return self._send_log(
            crud_operation='CREATE',
            client_id=client_id,
            agent_id=agent_id,
            after_value=after_value,
            correlation_id=correlation_id
        )
    
    def log_read(self, client_id: str, agent_id: str, 
                 correlation_id: Optional[str] = None) -> bool:
        """
        Log a READ operation.
        
        Args:
            client_id: ID of the client
            agent_id: ID of the agent performing the operation
            correlation_id: Optional correlation ID for request tracing
            
        Returns:
            True if log was sent successfully, False otherwise
        """
        return self._send_log(
            crud_operation='READ',
            client_id=client_id,
            agent_id=agent_id,
            correlation_id=correlation_id
        )
    
    def log_update(self, client_id: str, agent_id: str, attribute_name: str,
                   before_value: str, after_value: str,
                   correlation_id: Optional[str] = None) -> bool:
        """
        Log an UPDATE operation.
        
        Args:
            client_id: ID of the client
            agent_id: ID of the agent performing the operation
            attribute_name: Name of the attribute being updated (e.g., "First Name")
            before_value: Previous value (e.g., "LEE")
            after_value: New value (e.g., "TAN")
            correlation_id: Optional correlation ID for request tracing
            
        Returns:
            True if log was sent successfully, False otherwise
        """
        return self._send_log(
            crud_operation='UPDATE',
            client_id=client_id,
            agent_id=agent_id,
            attribute_name=attribute_name,
            before_value=before_value,
            after_value=after_value,
            correlation_id=correlation_id
        )
    
    def log_delete(self, client_id: str, agent_id: str, before_value: str,
                   correlation_id: Optional[str] = None) -> bool:
        """
        Log a DELETE operation.
        
        Args:
            client_id: ID of the client
            agent_id: ID of the agent performing the operation
            before_value: The value being deleted (e.g., "LEE|ABC")
            correlation_id: Optional correlation ID for request tracing
            
        Returns:
            True if log was sent successfully, False otherwise
        """
        return self._send_log(
            crud_operation='DELETE',
            client_id=client_id,
            agent_id=agent_id,
            before_value=before_value,
            correlation_id=correlation_id
        )
    
    def _send_log(self, crud_operation: str, client_id: str, agent_id: str,
                  attribute_name: Optional[str] = None,
                  before_value: Optional[str] = None,
                  after_value: Optional[str] = None,
                  correlation_id: Optional[str] = None) -> bool:
        """
        Internal method to send audit log message to SQS.
        
        Args:
            crud_operation: CRUD operation type
            client_id: Client ID
            agent_id: Agent ID
            attribute_name: Attribute name (for UPDATE operations)
            before_value: Previous value
            after_value: New value
            correlation_id: Correlation ID for tracing
            
        Returns:
            True if successful, False otherwise
        """
        try:
            # Create the audit log message
            message = {
                'crud_operation': crud_operation,
                'client_id': client_id,
                'agent_id': agent_id,
                'timestamp': datetime.utcnow().isoformat() + 'Z',  # ISO 8601 format
                'source_service': self.source_service,
                'correlation_id': correlation_id or str(uuid.uuid4())
            }
            
            # Add optional fields if provided
            if attribute_name:
                message['attribute_name'] = attribute_name
            if before_value:
                message['before_value'] = before_value
            if after_value:
                message['after_value'] = after_value
            
            # Send message to SQS
            response = self.sqs.send_message(
                QueueUrl=self.queue_url,
                MessageBody=json.dumps(message),
                MessageAttributes={
                    'operation_type': {
                        'StringValue': crud_operation,
                        'DataType': 'String'
                    },
                    'client_id': {
                        'StringValue': client_id,
                        'DataType': 'String'
                    },
                    'source_service': {
                        'StringValue': self.source_service,
                        'DataType': 'String'
                    }
                }
            )
            
            logger.info(f"Audit log sent successfully. MessageId: {response['MessageId']}")
            return True
            
        except ClientError as e:
            logger.error(f"Failed to send audit log to SQS: {e.response['Error']['Message']}")
            return False
        except Exception as e:
            logger.error(f"Unexpected error sending audit log: {str(e)}")
            return False


# Convenience function to create logger instance from environment variables
def create_audit_logger_from_env(source_service: str) -> AuditLogger:
    """
    Create an AuditLogger instance using environment variables.
    
    Expected environment variables:
    - AUDIT_LOG_QUEUE_URL: SQS queue URL for audit logs
    
    Args:
        source_service: Name of the service using this logger
        
    Returns:
        Configured AuditLogger instance
        
    Raises:
        ValueError: If required environment variables are missing
    """
    import os
    
    queue_url = os.environ.get('AUDIT_LOG_QUEUE_URL')
    if not queue_url:
        raise ValueError("AUDIT_LOG_QUEUE_URL environment variable is required")
    
    return AuditLogger(queue_url=queue_url, source_service=source_service)

File: services/shared/java/src/main/java/com/bankingbuddy/audit/AuditLogger.java
package com.bankingbuddy.audit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.DynamoDbException;
import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Mission-critical synchronous audit logger for Spring Boot applications.
 * Writes directly to DynamoDB for immediate consistency.
 * CRUD operations will fail if audit logging fails.
 */
@Component
public class AuditLogger {
    
    private static final Logger logger = LoggerFactory.getLogger(AuditLogger.class);
    
    private final DynamoDbClient dynamoDb;
    private final String tableName;
    private final String sourceService;
    private final long logRetentionDays;
    
    public AuditLogger(DynamoDbClient dynamoDb,
                      @Value("${audit.dynamodb.table.name}") String tableName,
                      @Value("${audit.source.service}") String sourceService,
                      @Value("${audit.log.retention.days:2555}") long logRetentionDays) {
        if (dynamoDb == null) {
            throw new NullPointerException("DynamoDB client cannot be null");
        }
        if (tableName == null) {
            throw new IllegalArgumentException("Table name cannot be null");
        }
        if (tableName.trim().isEmpty()) {
            throw new IllegalArgumentException("Table name cannot be empty");
        }
        if (sourceService == null) {
            throw new IllegalArgumentException("Source service cannot be null");
        }
        if (sourceService.trim().isEmpty()) {
            throw new IllegalArgumentException("Source service cannot be empty");
        }
        
        this.dynamoDb = dynamoDb;
        this.tableName = tableName;
        this.sourceService = sourceService;
        this.logRetentionDays = logRetentionDays;
    }
    
        /**
     * Log a CREATE operation.
     * @param clientId ID of the client
     * @param agentId ID of the agent
     * @param afterValue The created value (e.g., "LEE|ABC Street")
     * @throws AuditException if audit logging fails (causes transaction rollback)
     */
    public void logCreate(String clientId, String agentId, String afterValue) {
        writeToDynamoDB("CREATE", clientId, agentId, null, null, afterValue);
    }
    
    /**
     * Log a READ operation.
     * @param clientId ID of the client
     * @param agentId ID of the agent
     * @throws AuditException if audit logging fails (causes transaction rollback)
     */
    public void logRead(String clientId, String agentId) {
        writeToDynamoDB("READ", clientId, agentId, null, null, null);
    }
    
    /**
     * Log an UPDATE operation.
     * @param clientId ID of the client
     * @param agentId ID of the agent
     * @param attributeName Name of the attribute (e.g., "First Name")
     * @param beforeValue Previous value
     * @param afterValue New value
     * @throws AuditException if audit logging fails (causes transaction rollback)
     */
    public void logUpdate(String clientId, String agentId, String attributeName, 
                         String beforeValue, String afterValue) {
        writeToDynamoDB("UPDATE", clientId, agentId, attributeName, beforeValue, afterValue);
    }
    
    /**
     * Log a DELETE operation.
     * @param clientId ID of the client
     * @param agentId ID of the agent
     * @param beforeValue The deleted value (e.g., "LEE|ABC Street")
     * @throws AuditException if audit logging fails (causes transaction rollback)
     */
    public void logDelete(String clientId, String agentId, String beforeValue) {
        writeToDynamoDB("DELETE", clientId, agentId, null, beforeValue, null);
    }
    
    /**
     * Synchronously write audit log directly to DynamoDB.
     * Mission-critical: throws exception if write fails.
     */
    private void writeToDynamoDB(String operation, String clientId, String agentId,
                                String attributeName, String beforeValue, String afterValue) {
        // Validate required fields
        if (clientId == null || clientId.trim().isEmpty()) {
            throw new IllegalArgumentException("Client ID cannot be null or empty");
        }
        if (agentId == null || agentId.trim().isEmpty()) {
            throw new IllegalArgumentException("Agent ID cannot be null or empty");
        }
        if (operation == null || operation.trim().isEmpty()) {
            throw new IllegalArgumentException("Operation cannot be null or empty");
        }
        
        try {
            Instant now = Instant.now();
            String timestamp = now.toString();
            
            // Calculate TTL for automatic deletion
            long ttlEpochSeconds = now.plus(logRetentionDays, ChronoUnit.DAYS).getEpochSecond();
            
            // Build DynamoDB item with required fields
            Map<String, AttributeValue> item = new HashMap<>();
            item.put("client_id", AttributeValue.builder().s(clientId.trim()).build());
            item.put("timestamp", AttributeValue.builder().s(timestamp).build());
            item.put("log_id", AttributeValue.builder().s(UUID.randomUUID().toString()).build());
            item.put("crud_operation", AttributeValue.builder().s(operation.trim()).build());
            item.put("agent_id", AttributeValue.builder().s(agentId.trim()).build());
            item.put("source_service", AttributeValue.builder().s(sourceService).build());
            item.put("ttl", AttributeValue.builder().n(String.valueOf(ttlEpochSeconds)).build());
            
            // Optional fields
            if (attributeName != null && !attributeName.isEmpty()) {
                item.put("attribute_name", AttributeValue.builder().s(attributeName).build());
            }
            if (beforeValue != null && !beforeValue.isEmpty()) {
                item.put("before_value", AttributeValue.builder().s(beforeValue).build());
            }
            if (afterValue != null && !afterValue.isEmpty()) {
                item.put("after_value", AttributeValue.builder().s(afterValue).build());
            }
            
            // Synchronous write to DynamoDB
            PutItemRequest request = PutItemRequest.builder()
                    .tableName(tableName)
                    .item(item)
                    .build();
            
            dynamoDb.putItem(request);
            
            logger.info("Audit log written successfully: {} for client {}", operation, clientId);
            
        } catch (DynamoDbException e) {
            logger.error("DynamoDB audit logging failed: {}", e.getMessage(), e);
            throw new AuditException("Mission-critical audit logging failed: " + e.getMessage(), e);
        } catch (Exception e) {
            logger.error("Unexpected audit logging error: {}", e.getMessage(), e);
            throw new AuditException("Mission-critical audit logging failed: " + e.getMessage(), e);
        }
    }
    
    /**
     * Exception thrown when audit logging fails.
     * Causes transaction rollback in mission-critical operations.
     */
    public static class AuditException extends RuntimeException {
        public AuditException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}